k4Reco/GaudiLumiCalClusterer/include/GaudiLumiCalClusterer.h

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2020-2024 Key4hep-Project.
+ *
+ * This file is part of Key4hep.
+ * See https://key4hep.github.io/key4hep-doc/ for further info.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef K4RECO_GAUDILUMICALCLUSTERER_H
+#define K4RECO_GAUDILUMICALCLUSTERER_H
+
+#include "GlobalMethodsClass.h"
+#include "LumiCalClusterer.h"
+// #include "OutputManagerClass.h"
+
+#include <edm4hep/CalorimeterHitCollection.h>
+#include <edm4hep/ClusterCollection.h>
+#include <edm4hep/ReconstructedParticleCollection.h>
+#include <edm4hep/SimCalorimeterHitCollection.h>
+
+#include <Gaudi/Property.h>
+
+#include <k4FWCore/Transformer.h>
+#include <k4Interface/IGeoSvc.h>
+
+#include <map>
+#include <string>
+#include <tuple>
+#include <vector>
+
+class LCCluster;
+class ClusterClass;
+
+typedef std::map<int, ClusterClass*> MapIntPClusterClass;
+typedef std::map<int, std::vector<int>> MapIntVInt;
+
+struct GaudiLumiCalClusterer
+    : k4FWCore::MultiTransformer<std::tuple<edm4hep::CalorimeterHitCollection, edm4hep::ClusterCollection,
+                                            edm4hep::ReconstructedParticleCollection>(
+          const edm4hep::SimCalorimeterHitCollection&)> {
+
+public:
+  GaudiLumiCalClusterer(const std::string& name, ISvcLocator* svcLoc);
+
+  StatusCode initialize() override;
+
+  std::tuple<edm4hep::CalorimeterHitCollection, edm4hep::ClusterCollection, edm4hep::ReconstructedParticleCollection>
+  operator()(const edm4hep::SimCalorimeterHitCollection&) const override;
+
+  // Processor Parameters
+  std::string LumiInColName;
+  std::string LumiOutColName = "LumiCalHits";
+  double m_BeamCrossingAngle;
+
+  GlobalMethodsClass gmc;
+  mutable LumiCalClustererClass LumiCalClusterer;
```

Comment:
Is this thread-safe? I don't see anything that would tell the scheduler that this `MultiTransformer` is not re-entrant, so there is some potential for threading issues here.

Reply:
Nope so I added an `isReEntrant` function.